modules/stochastic_tools/src/reporters/DirectPerturbationReporter.C

```diff
@@ -0,0 +1,251 @@
+    const bool right_i_in_owned_range =
+        _sampler.getLocalRowBegin() <= right_i && right_i < _sampler.getLocalRowEnd();
+
+    if (left_i_in_owned_range || right_i_in_owned_range)
+    {
+      const dof_id_type copy_i = left_i_in_owned_range ? left_i - offset : right_i - offset;
+      this->_state.value()[param_i] = initializeSensitivity(_data[copy_i]);
+    }
+
+    // We add the contribution from one side
+    if (left_i_in_owned_range)
+      addSensitivityConstribution(this->_state.value()[param_i], _data[left_i - offset], interval);
+
+    // We add the contribution from the other side
+    if (right_i_in_owned_range)
+      addSensitivityConstribution(
+          this->_state.value()[param_i], _data[right_i - offset], -interval);
+  }
+
+  // We gather the contributions across all processors
+  this->vectorSum();
+}
+
+template <typename DataType>
+void
+DirectPerturbationReporterContext<DataType>::addSensitivityConstribution(DataType & add_to,
+                                                                         const DataType & to_add,
+                                                                         const Real interval) const
+{
+  // DataType is a numeric type that we can sum (excluding bool)
+  if constexpr (std::is_arithmetic<DataType>::value && !std::is_same<DataType, bool>::value)
+  {
+    add_to += to_add / interval;
+    return;
+  }
+  // DataType is a vector type
+  else if constexpr (is_std_vector<DataType>::value)
+  {
+    // It can still be anything in the vector elements, so we will check this later
+    using VectorValueType = typename DataType::value_type;
+
+    mooseAssert(add_to.size() == to_add.size(), "The vectors for summation have different sizes!");
+
+    // Check if the vector elements are of a numeric type
+    if constexpr (std::is_arithmetic<VectorValueType>::value &&
+                  !std::is_same<VectorValueType, bool>::value)
+    {
+      for (const auto index : index_range(add_to))
+        add_to[index] += to_add[index] / interval;
+      return;
+    }
+    // Check if the vector elements are also vectors
+    else if constexpr (is_std_vector<VectorValueType>::value)
+    {
+      // This is as deep as we will go for now
+      using InnerValueType = typename VectorValueType::value_type;
+
+      // Check if the inner vector elements are of a numeric type
+      if constexpr (std::is_arithmetic<InnerValueType>::value &&
+                    !std::is_same<InnerValueType, bool>::value)
+      {
+        // Iterate over each inner vector in the outer vector
+        for (auto & inner_index : index_range(add_to))
+        {
+          mooseAssert(add_to[inner_index].size() == to_add[inner_index].size(),
+                      "The vectors for summation have different sizes!");
+          for (const auto index : index_range(add_to[inner_index]))
+            add_to[inner_index][index] += to_add[inner_index][index] / interval;
```

Comment:
Need more of these in the nested if conditions

Reply:
by these I meant
```
    static_assert(Moose::always_false<DataType>,
                  "Sensitivity coefficient computation is not implemented for the given type!");
```

Reply:
Yeah addded, will push soon! Very nice catch!